### PR TITLE
Add department-specific radio channels

### PR DESCRIPTION
**Pull request by Partmedia**
## About the PR 
This PR adds working department-specific radio channels, while minimizing damage to the current codebase. It is expected that a future refactor will clean this up a bit.

ChatSystem now has a RadioPrefix() method that recognizes department-specific channels (e.g. ":e" and ":m") in addition to the
global channel (";"). It strips the prefix from the message and assigns messages an integer representing the destination channel, if any.

IListen and IRadio now accept optional 'channel' arguments with this channel in mind.

The ugly is that the integer channel number is hard-coded and also shows up in chat.

Comms are not modeled at this time. You cannot break comms (yet).

All headsets have channels soldered into them. You cannot change encryption keys to hop on new channels. Steal a headset instead.

This is currently server-side only. To use a department specific channel, manually add the prefix (e.g. ":e") to "Local" chat.

**Screenshots**

**Changelog**

:cl: notafet
- add: Headsets can now use department-specific radio channels.

### CODE REVIEW

File: Content.Server/Chat/ChatSystem.cs [hunk 1]
```diff
@@ -329,19 +331,76 @@ private IEnumerable<INetChannel> GetDeadChatClients()
...
+                    _popup.PopupEntity(Loc.GetString("chat-manager-no-such-channel"), source, Filter.Entities(source));
+                    chan = -1;
+                    break;
+                }
+
+                // Strip message prefix.
+                var parts = message.Split(' ').ToList();
+                parts.RemoveAt(0);
+                message = string.Join(" ", parts);
+            }
```
Comment (metalgearsloth) [inline Content.Server/Chat/ChatSystem.cs:397]: These should be done in prototypes as using it is cumbersome using this in prototypes as is.

Also don't do console.writelines, use the logger.

**Comment by metalgearsloth** [comment]
Also I think @LetterN was working on radios.

File: Content.Server/Chat/ChatSystem.cs (hunk 1 above)
Comment (Partmedia) [reply Content.Server/Chat/ChatSystem.cs:396]: Console.WriteLines() was left from debugging, oops! It has now been removed.

**Comment by Partmedia** [comment]
I checked through issues and PR's and didn't find any current progress on radios. I should have asked on Discord.

I see this PR as a stop-gap measure to get radios into gameplay, without finishing other aspects like changing encryption keys and fully-modeled comms. This adds "working" albeit not quite complete radios, so that, e.g. nukies can start coordinating.

@LetterN I'd be more than happy to help you with the finished radio system, if you would like me to!

File: Content.Server/Chat/ChatSystem.cs [hunk 2]
```diff
@@ -329,19 +334,45 @@ private IEnumerable<INetChannel> GetDeadChatClients()
             .Select(p => p.ConnectedClient);
     }
 
-    private string SanitizeMessageCapital(EntityUid source, string message)
+    private (string, int) RadioPrefix(EntityUid source, string message)
```
Comment (metalgearsloth) [inline Content.Server/Chat/ChatSystem.cs:337]: This should be returning the prototype string, not the channel int. Also comments.

File: Content.Server/Chat/ChatSystem.cs [hunk 3]
```diff
@@ -89,6 +92,8 @@ private void OnLoocEnabledChanged(bool val)
 
         bool shouldCapitalize = (desiredType != InGameICChatType.Emote);
 
+        int channel;
+        (message, channel) = RadioPrefix(source, message);
         message = SanitizeInGameICMessage(source, message, out var emoteStr, shouldCapitalize);
```
Comment (metalgearsloth) [inline Content.Server/Chat/ChatSystem.cs:97]: Is there a reason to get the radio prefix before checking if the entity can speak?

File: Content.Server/Headset/HeadsetComponent.cs [hunk 4]
```diff
@@ -41,6 +44,17 @@ protected override void Initialize()
...
+        private List<int> getChannels()
+        {
+            List<int> ret = new List<int>();
+            foreach (string channelId in _channels)
+            {
+                if (IoCManager.Resolve<IPrototypeManager>().TryIndex(channelId, out RadioChannelPrototype? proto))
+                    ret.Add(proto.Channel);
+            }
+            return ret;
+        }
```
Comment (metalgearsloth) [inline Content.Server/Headset/HeadsetComponent.cs:56]: No logic in components please, also C# styleguide for method name (GetChannels)

File: Content.Server/Headset/HeadsetComponent.cs [hunk 5]
```diff
@@ -19,18 +22,18 @@ public sealed class HeadsetComponent : Component, IListen, IRadio
...
 
-        [DataField("channels")]
-        private List<int> _channels = new(){1459};
+        [DataField("channels", customTypeSerializer:typeof(PrototypeIdListSerializer<RadioChannelPrototype>))]
+        private List<String> _channels = new(){"broadcast"};
 
         [ViewVariables(VVAccess.ReadWrite)]
         [DataField("broadcastChannel")]
-        public int BroadcastFrequency { get; set; } = 1459;
+        public int BroadcastFrequency { get; set; } = 0;
```
Comment (metalgearsloth) [inline Content.Server/Headset/HeadsetComponent.cs:30]: This seems not necessary?

File: Content.Shared/Radio/RadioChannelPrototype.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Radio
+{
+    [Prototype("radioChannel")]
+    public sealed class RadioChannelPrototype : IPrototype
+    {
+        [ViewVariables] [DataField("name")] public string Name { get; private set; } = string.Empty;
+        [ViewVariables] [DataField("keycode")] public char KeyCode { get; private set; } = '\0';
+        [ViewVariables] [DataField("channel")] public int Channel { get; private set; } = 0;
```
Comment (metalgearsloth) [inline Content.Shared/Radio/RadioChannelPrototype.cs:10]: Comments, e.g. "KeyCode is the prefix used to determine what channel a message corresponds to).

File: Resources/Prototypes/Entities/Clothing/Ears/headsets.yml [hunk 7]
```diff
@@ -19,6 +19,8 @@
   name: cargo headset
   description: A headset used by the quartermaster and his slaves.
   components:
+  - type: Headset
+    channels: [broadcast, supply]
   - type: Sprite
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Clothing/Ears/headsets.yml:24]: IMO it's much nicer to do these as a list, like so:

```suggestion
  - type: Headset
    channels:
    - broadcast
    - supply
  - type: Sprite
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
There's still some logic tied to using ints as channels which shouldn't be there, the channels should 100% be prototypes and the int should just be cosmetic. I think it's like 80% of the way there.

**Comment by metalgearsloth** [comment]
I'll come back to this when I get a chance.

**Pull request closed without merging**